src/core/border_router/routing_manager.cpp

```diff
@@ -221,14 +240,70 @@ Error RoutingManager::LoadOrGenerateRandomBrUlaPrefix(void)
 }
 
 #if OPENTHREAD_CONFIG_BORDER_ROUTING_NAT64_ENABLE
+const otIp4Address RoutingManager::kWellKnownIpv4OnlyAddress1 = {{{192, 0, 0, 170}}};
+const otIp4Address RoutingManager::kWellKnownIpv4OnlyAddress2 = {{{192, 0, 0, 171}}};
+
+void RoutingManager::DiscoverInfraIfNat64Prefix(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsInitialized() && mInfraIf.IsRunning(), error = kErrorInvalidState);
+
+    LogInfo("Discovering infrastructure NAT64 prefix on %s", mInfraIf.ToString().AsCString());
+    error = mInfraIf.DiscoverNat64Prefixes();
+
+exit:
+    if (error != kErrorNone)
+    {
+        LogWarn("Failed to request infrastructure NAT64 prefix on %s: %s", mInfraIf.ToString().AsCString(),
+                ErrorToString(error));
+    }
+}
+
+void RoutingManager::DiscoverInfraIfNat64PrefixesDone(const Ip6::Address *aAddresses, uint8_t aNumAddresses)
+{
+    Ip6::Prefix   prefix;
+    const uint8_t prefixLength[] = {32, 40, 48, 56, 64, 96};
+
+    prefix.Clear();
+    for (uint8_t i = 0; aAddresses != nullptr && i < aNumAddresses && !prefix.IsValidNat64(); i++)
+    {
+        for (uint8_t length : prefixLength)
+        {
+            Ip4::Address ip4Address;
+
+            ip4Address.ExtractFromIp6Address(length, aAddresses[i]);
+            if (memcmp(ip4Address.GetBytes(), kWellKnownIpv4OnlyAddress1.mFields.m8, OT_IP4_ADDRESS_SIZE) == 0 ||
+                memcmp(ip4Address.GetBytes(), kWellKnownIpv4OnlyAddress2.mFields.m8, OT_IP4_ADDRESS_SIZE) == 0)
+            {
+                aAddresses->GetPrefix(length, prefix);
+                break;
```

Comment:
I can think of a corner case here:

What if the prefix happens to contain the same byte pattern  as one of the well-known IPv4 addresses. Then in the code here we may treat the prefix as a shorter one (since the address would match at a shorter length).

May be better if we change the prefix length array to start with longer prefix lengths first?

Reply:
I searched a bit and found that there seems to be a solution mentioned in [RFC-7050 section 3](https://datatracker.ietf.org/doc/html/rfc7050#section-3) about this:

> A node MUST look through all of the received AAAA resource records to
>    collect one or more Pref64::/n.  The Pref64::/n list might include
>    the Well-Known Prefix 64:ff9b::/96 [[RFC6052](https://datatracker.ietf.org/doc/html/rfc6052)] or one or more Network-
>    Specific Prefixes.  In the case of NSPs, the node SHALL determine the
>    used address format by searching the received IPv6 addresses for the
>    WKN's well-known IPv4 addresses.  The node SHALL assume the well-
>    known IPv4 addresses might be found at the locations specified by
>    [[RFC6052], Section 2.2](https://datatracker.ietf.org/doc/html/rfc6052#section-2.2).  **The node MUST check on octet boundaries to
>    ensure a 32-bit well-known IPv4 address value is present only once in
>    an IPv6 address.  In case another instance of the value is found
>    inside the IPv6 address, the node SHALL repeat the search with the
>    other well-known IPv4 address.**

Seems like we need to ensure that we see the same well-known address only once and if we can see get the same IPv4 address with a different NAT64 prefix lengths then we should skip this and/or use with other well-known address.... I guess that may be why there are two well-kwon addresses :)

My be good to implement this logic here.

Reply:
Thanks for pointing out this case! I added a TODO here and will implement the logic in a follow-up PR.

Reply:
Let's add it in this PR, so we can handle this case corrrectly.
I think something like this would work:
```c++
            if (ip4Address == AsCoreType(&kWellKnownIpv4OnlyAddress1) ||
                ip4Address == AsCoreType(&kWellKnownIpv4OnlyAddress2))
            {
                // We check that the well-known IPv4 address is
                // present only once in the IPv6 address. In case
                // another instance of it is found for another prefix
                // length, we ignore this address and check other
                // addresses with other well-known IPv4 address.
                // (per RFC 7050 section 3).

                bool foundDuplicate = false;

                for (uint8_t dupLength : kValidNat64PrefixLength)
                {
                    Ip4::Address dupIp4Address;

                    if (dupLength == length)
                    {
                        continue;
                    }

                    dupIp4Address.ExtractFromIp6Address(dupLength, aAddresses[i]));

                    if (dupIp4Address == ip4Address)
                    {
                        foundDuplicate = true;
                        break;
                    }
                }

                if (!foundDuplicate)
                {
                    aAddresses->GetPrefix(length, prefix);
                    break;
                }
            }
```

Reply:
Thanks, Abtin! Processing the received IPv6 addresses is changed to be in ot::Posix::InfraNetif, and the logic is also updated there to notify OT the discovered NAT64 prefix. In this way the platform API can better adapt with the Android platform.

Reply:
Sounds good to me. I agree makes the API and interaction with platform simpler.
We can do similar code on the platform side.